README
NAME
    Gitalist - A modern git web viewer

SYNOPSIS
        perl script/gitalist_server.pl --repo_dir /home/me/code/git

INSTALL
    As Gitalist follows the usual Perl module format the usual approach for
    installation should work, e.g.:

        perl Makefile.PL
        make
        make test
        make install

    or

        cpan -i Gitalist

    You can also check Gitalist out from its git repository and run it, in
    this case you'll additionally need the author modules, but no
    configuration will be needed as it will default to looking for
    repositories the directory above the checkout.

DESCRIPTION
    Gitalist is a web frontend for git repositories based on gitweb.cgi
    <https://git.wiki.kernel.org/index.php/Gitweb> and backed by Catalyst.

  History
    This project started off as an attempt to port *gitweb.cgi* to a
    Catalyst app in a piecemeal fashion. As it turns out, thanks largely to
    Florian Ragwitz's earlier effort, it was easier to use *gitweb.cgi* as a
    template for building a new Catalyst application.

GETTING GITALIST
    You can install Gitalist from CPAN in the usual way:

        cpan -i Gitalist

    Alternatively, you can get Gitalist using git.

    The canonical repository for the master branch is:

        git://git.shadowcat.co.uk/catagits/Gitalist.git

    Gitalist is also mirrored to GitHub at
    <https://github.com/broquaint/Gitalist>, and a number of people have
    active forks with branches and/or new features in the master branch.

BOOTSTRAPPING
    As of 0.002001 Gitalist can now be bootstrapped to run out of its own
    directory by installing its prerequisites locally with the help of
    local::lib. So instead of installing the prerequisites to the system
    path with CPAN they are installed under the Gitalist directory.

    To do this clone Gitalist from the Shadowcat repository mentioned above
    or grab a snapshot from broquaint's GitHub repository:

        https://github.com/broquaint/Gitalist/downloads

    With the source acquired and unpacked run the following from within the
    Gitalist directory:

        perl script/bootstrap.pl

    This will install the necessary modules for the build process which in
    turn installs the prerequisites locally.

    NB: The relevant bootstrap scripts aren't available in the CPAN dist as
    the bootstrap scripts should not be installed.

INITIAL CONFIGURATION
    Gitalist is configured using Catalyst::Plugin::Configloader. The
    supplied sample configuration is in Config::General format, however it
    is possible to configure Gitalist using other config file formats (such
    as YAML) if you prefer.

  WHEN CHECKING GITALIST OUT OF GIT
    Gitalist from git includes a minimal "gitalist_local.conf", which sets
    the repository directory to one directory higher than the Gitalist
    repository.

    This means that if you check Gitalist out next to your other git
    checkouts, then starting the demo server needs no parameters at all:

        Gitalist [master]$ perl script/gitalist_server.pl
        You can connect to your server at http://localhost:3000

  FOR CPAN INSTALLS
    Gitalist can be supplied with a config file by setting the
    "GITALIST_CONFIG" environment variable to point to a configuration file.

    If you install Gitalist from CPAN, a default configuration is installed
    along with gitalist, which is complete except for a repository
    directory. You can get a copy of this configuration by running:

      cp `perl -Ilib -MGitalist -e'print Gitalist->path_to("gitalist.conf")'` gitalist.conf

    You can then edit this configuration, adding a "repo_dir" path and
    customising other settings as desired.

    You can then start the Gitalist demo server by setting
    "GITALIST_CONFIG". For example:

        GITALIST_CONFIG=/usr/local/etc/gitalist.conf gitalist_server.pl

    Alternatively, if you only want to set a repository directory and are
    otherwise happy with the default configuration, then you can set the
    "GITALIST_REPO_DIR" environment variable, or pass the "--repo_dir" flag
    to any of the scripts.

        GITALIST_REPO_DIR=/home/myuser/code/git gitalist_server.pl
        gitalist_server.pl --repo_dir home/myuser/code/git

    The "GITALIST_REPO_DIR" environment variable will override the
    repository directory set in configuration, and will itself be overridden
    by he "--repo_dir" flag.

RUNNING
    Once you have followed the instructions above to install and configure
    Gitalist, you may want to run it in a more production facing environment
    than using the single threaded developement server.

    The recommended deployment method for Gitalist is FastCGI, although
    Gitalist can also be run under mod_perl <https://perl.apache.org/> or as
    pure Perl with Catalyst::Engine::PreFork.

    Assuming that you have installed Gitalist's dependencies into a
    local::lib, and you are running from a git checkout, adding a trivial
    FCGI script as "script/gitalist.fcgi" (this file is specifically in
    ".gitignore" so you can have your own copy):

        #!/bin/sh
        exec /home/t0m/public_html/Gitalist/script/gitalist_fastcgi.pl

    This example can be seen live here:

        http://example.gitalist.com

  Plack
    If you would like to run Gitalist under Plack then one need only make
    use of plackup and the ".psgi" found under "scripts/":

        plackup script/gitalist_app.psgi

CONFIGURATION
    The Gitalist config is loaded with Catalyst::Plugin::ConfigLoader and
    the available config options are:

    no_gravatars
        If true don't display gravatars. May be desirable if you are worried
        about leaking repo names via the browser referer headers.

  Model::CollectionOfRepos
    repo_dir
        A directory containing the directories to show.

    repos
        A list of directories containing repositories to show.

    search_recursively
        A boolean indicating whether to always search recursively for
        repositories within "repo_dir".

    whitelist
        Path a file containing a list of repositories that can be shown.
        Each line in the file will represent the name of a repo to show e.g

          Gitalist
          some-bare-repo.git

        This is compatible with "gitweb"'s "projects.list".

    export_ok
        If provided every must contain a file of the same name to be
        visible. This is similar to "gitweb"'s functionality.

    class
        If you want a different way of surfacing repositories you can use
        your own model (i.e something that composes
        Gitalist::Git::CollectionOfRepositories) and specify the class name
        with this config option.

    args
        Any additional arguments to be passed into the Model constructor,
        only of use when used in conjunction with "class" (see above).

  paging
    log The number of commits to show in the summary, shortlog and longlog
        views.

  FASTCGI
    Running Gitalist in FastCGI mode requires a webserver with FastCGI
    support (such as apache with mod_fcgi
    <http://www.fastcgi.com/drupal/node/3> or mod_fcgid
    <https://httpd.apache.org/mod_fcgid/>). Below is a sample configuration
    using Apache2 with mod_fcgid in a dynamic configuration (as opposed to
    static or standalone mode). More information on these modes and their
    configuration can be found at "Standalone server mode" in
    Catalyst::Engine::FastCGI.

    In Apache's mime.conf, add "AddHandler fcgid-script .fcgi" (or
    "AddHandler fastcgi-script .fcgi" for mod_fcgi).

    And a quick VirtualHost configuration:

        <VirtualHost *:80>
            ServerName gitalist.yourdomain.com
            DocumentRoot /path/to/gitalist.fcgi
            <Directory "/path/to/gitalist.fcgi">
                AllowOverride all
                Options +ExecCGI -MultiViews +SymLinksIfOwnerMatch
                Order allow,deny
                Allow from all
            </Directory>

            # Tell Apache this is a FastCGI application
            <Files gitalist.fcgi>
                #change the below to fastcgi-script if using mod_fcgi
                SetHandler fcgid-script
            </Files>
        </VirtualHost>

    Now to access your Gitalist instance, you'll go to
    "gitalist.yourdomain.com/gitalist.fcgi/" (do not forget that trailing
    "/"). If you'd like a different URL, of course, you'll likely want to
    use mod_rewrite <https://httpd.apache.org/docs/mod/mod_rewrite.html> or
    equivalent.

    If you find the need to do some troubleshooting, you can call
    "http://url_to_gitalist.fcgi?dump_info=1" and/or add export
    "GITALIST_DEBUG=1" to the top of your gitalist.fcgi file (just below the
    shebang line).

  Apache config
    Apache will refuse %2F in Gitalist URLs unless configured otherwise.
    Make sure "AllowEncodedSlashes On" is in your httpd.conf file in order
    for this to run smoothly.

    To have the static content served statically by Apache, instead of
    Gitalist, then add something like following line to your httpd.conf:

        Alias /static  /usr/local/share/perl/5.10.1/Gitalist/root/static

CONTRIBUTING
    Patches are welcome, please feel free to fork on github and send pull
    requests, send patches from git format-patch to the bug tracker, or host
    your own copy of gitalist somewhere and ask us to pull from it.

SUPPORT
    Gitalist has an active irc community in "#gitalist" on irc.perl.org,
    please feel free to stop by and ask questions, report bugs or
    installation issues or generally for a chat about where we plan to go
    with the project.

SEE ALSO
    Gitalist::Controller::Root

    Git::Gitalist::Repository

    Catalyst

AUTHORS AND COPYRIGHT
      Catalyst application:
        © 2009 Venda Ltd and Dan Brook <broq@cpan.org>
        © 2009, Tom Doran <bobtfish@bobtfish.net>
        © 2009, Zac Stevens <zts@cryptocracy.com>

      Original gitweb.cgi from which this was derived:
        © 2005-2006, Kay Sievers <kay.sievers@vrfy.org>
        © 2005, Christian Gierke

      Model based on http://github.com/rafl/gitweb
        © 2008, Florian Ragwitz

LICENSE
    Licensed under GNU GPL v2



Changes
This file documents the revision history for Perl extension Gitalist.

0.005001 2014-02-11
  - Stable version of dev release, now depends on latest Git::Gitalist.

0.005000_01 2014-02-05
  - Depend on Git::Gitalist to provide git logic.

0.004004 2014-01-25
  - Move away from Class::MOP::load_class (Zoffix Znet)

0.004003 2013-09-04
  - Handle non-trivial tags when displaying associated tree.

0.004002 2013-04-01
  - Update gitalist_app.psgi with working copy.
  - Fix the bare.git test repo.
  - The branch selector now sends you to the correct url (Dennis Kaarsemaker)
  - Massive tree action speedup (Dennis Kaarsemaker)
  - Have line numbers in blob displays (Dennis Kaarsemaker)
  - Syntax higlighting improvements (Dennis Kaarsemaker)

0.004001 2013-03-15
  - Depend on new Catalyst::Runtime and drop use of Catalyst::Controller::ActionRole
  - Stop calling tags branches (Dennis Kaarsemaker)
  - Stop breaking trees when submodules are encountered (Dennis Kaarsemaker)
  - Correctly handle utf-8 encoded .git/description files (Dennis Kaarsemaker)
  - Controller::ActionRole is dead (Tomas Doran)
  - Add gitalist.psgi (Tomas Doran)

0.003009 2012-03-18
  - Don't use MooseX::Types::ISO8601 to serialize DateTime objects as its
    coercion function is now stricter than the object provided.
  - Make tests pass with more recent versions of Catalyst.

0.003008 2012-01-29
  - Now with Changes.

0.003007 2012-01-28
  - Allow the model to be user defined,
  - Use new Catalyst features.
  - Make repo_dir work consistently with GITALIST_REPO_DIR (Tomas Doran).
  - Add the Vhost collection of repositories (Tomas Doran).

0.003006 2011-10-18
  - Fix lies in docs, make config more obvious (Tomas Doran)
  - Avoid "Use of qw(...) as parentheses is deprecated" warning with
    perl-5.14 (Gisle Aas)

0.003005 2011-07-13
  - Provide option to disable gravatars.
  - Tweaks to make Gitalist fit for debian (Jonas Genannt).
  - Fixes to enable Gitalist to run on Win32 (Christian Walde).
  - Fix content-type of core.css handler (Christian Walde).

0.003004 2011-06-07
  - Fix dependency typo in Makefile.PL.

0.003003 2011-06-06
  - Fixed ownership assertions in t/02git_object.t (Christian Walde).
  - Change default local config to not search recursively.
  - Don't attempt to syntax highlight anything big.
  - Fix for infinite loop bug in blame.

0.003002 2011-05-23
  - Add JSON serialization and expose via .json (Tomas Doran).
  - Provide .psgi for Plack goodness (Francoise Dehinbo).

0.003001 2011-03-13
  - Reinstate FromDirectory in model.
  - Drop unneeded deps in Makefile.PL.
  - Add support for multiple repos with the same name (Jose Luis Martinez).
  - Make paging work for history (Lian Wan Situ),
  - Respect paging.log config settings.

0.002009 2011-03-06
  - Document the available configuration options.
  - Provide switch in config for forcing recursive repo searching.
  - Assume UTF8 encoded blobs.

0.002008 2011-02-26
  - Add the ability to provide a whitelist of repos.
  - Add support for export-ok config option.
  - Ignore ancient/malformed tags.
  - Use the correct perl binary we installed against (Tom Hukins)

0.002007 2011-02-19
  - Add support for recursively searching for git repos (Dipesh Patel)
  - Cleaned up model code in Gitalist::Git:: (Zachary Stevens)
  - Improve POD mark-up, some typos (Lars Dɪᴇᴄᴋᴏᴡ 迪拉斯)
  - Regenerated README and fixed tests for bootstrapped checkouts.
  - Add Test::Exception dep.

0.002006 2010-10-17
  - Added notes regarding slash encoding in Catalyst
  - Fixed typo in core.css
  - Updated formatting in Gitalist.pm
  - Modified README with more details on FCGI deployment with Apache
  - Fix broken env includes in script/gitalist*.pl.

0.002005 2010-08-13
  - Limit the number of branches/tags listed by default.

0.002004 2010-08-02
  - Fix misnamed variable which was giving a syntax error in
    Gitalist::Controller::Fragment::Ref.
  - Fix blame view to display syntax highlighted blobs correctly
    instead of outputting raw HTML.

0.002003 2010-08-02
  - Fix for distribution not including script/env causing Makefile.PL
    and scripts to die.

0.002002 2010-08-01
  - Fix dist packaging issues.

0.002001 2010-07-03
  - Fix for loadCommitInfo.
  - Fix tests to work with local-libbery.
  - Correct example link and FCGI script.
  - Bumping version and updated Changes appropriately.
  - Throttle the commit info loading.
  - Update bootstrapping for local::lib 1.006000.
  - Fixed author_* barewords in Makefile.PL.
  - Handle merges in the diff view.
  - Add script/{bootstrap.pl,env} to MANIFEST.SKIP.

0.001008 2010-06-03
  - Add a requires_external_bin for the git command so we won't try running
    tests unless it's present.
  - Remove MooseX::MultiMethods as it has been broken by a recent Moose
    change. The new code is smaller and will be faster.
  - Add diagnostics to the tests to print out the version of git that
    people are using.
  - Make tests which commonly fail with older/different versions of git
    fail with the output of their git command to make debugging the tests
    from automated reports possible.

0.001007 2010-05-15
  - Set the use_request_uri_for_path configuration variable to be compatible
    with Catalyst 5.80024.

0.001006 2010-05-13
  - Fix Catalyst::Action::REST dependency (there wasn't one previously)
  - Depend on latest Catalyst::View::TT due to a previous bad version
    breaking things.
  - Add render_die => 1 in the View::TT config as we do want real exceptions
    when we call ->render ourselves.

0.001005 2010-05-11
  - Depend on latest MX::Declare to fix issues for people who upgrade
    MooseX::Types::Structured to a newer version.
  - Add some padding below, so file icons are not chopped off
  - Updated favicon to cope with dark grey backgrounds

0.001004 2010-05-06
  - Remove unneeded dependencies:
  - XML::OPML::SimpleGen
  - XML::Atom
  - XML::RSS
  - Fixed author_* barewords in Makefile.PL so that users checking out from

lib/Gitalist.pm
package Gitalist;

BEGIN { require 5.008006; }

use Moose;

extends 'Catalyst';

use Git::Gitalist::Util;

use Catalyst::Runtime 5.90006;
use Catalyst qw/
                ConfigLoader
                Unicode::Encoding
                Static::Simple
                StackTrace
                SubRequest
/;

use namespace::autoclean;

our $VERSION = '0.005000';
$VERSION = eval $VERSION;

__PACKAGE__->config(
    name          => 'Gitalist',
    default_view  => 'Default',
    default_model => 'CollectionOfRepos',
    use_request_uri_for_path => 1,
    disable_component_resolution_regex_fallback => 1,
);

__PACKAGE__->setup();

after prepare_path => sub {
    my ($ctx) = @_;
    my $path = $ctx->req->uri->path;
    if ($ctx->req->param('a')) {
        $ctx->req->uri->path("/legacy$path");
    }

    if($path =~ s/[.]json$// && $ctx->req->content_type eq 'application/json') {
        $ctx->req->uri->path($path);
    }
};

around uri_for => sub {
  my ($orig, $c) = (shift, shift);
  my $uri = $c->$orig(@_);
  $$uri =~ tr[&][;] if defined $uri;
  return $uri;
};

sub uri_with {
  my ($self, @args) = @_;
  my $uri = $self->request->uri_with(@args);
  # Wow this awful.
  $uri =~ s[/fragment\b][];
  return $uri;
}

after setup_finalize => sub {
    # At app startup, ensure we can find a git binary, rather than
    # lazily breaking later at request time.
    Git::Gitalist::Util::git_executable_path()
      or die <<ERR;
No git executable found in PATH.
Please specify which git executable to use in gitalist.conf or ensure that
a git executable is reachable in the PATH environment variable.
ERR
};

1;

__END__

=encoding UTF-8

=head1 NAME

Gitalist - A modern git web viewer

=head1 SYNOPSIS

    perl script/gitalist_server.pl --repo_dir /home/me/code/git

=head1 INSTALL

As Gitalist follows the usual Perl module format the usual approach
for installation should work, e.g.:

    perl Makefile.PL
    make
    make test
    make install

or

    cpan -i Gitalist

You can also L<check Gitalist out from its git repository|/"GETTING GITALIST">
and run it, in this case you'll additionally need the author modules,
but no configuration will be needed as it will default to looking
for repositories the directory above the checkout.

=head1 DESCRIPTION

Gitalist is a web frontend for git repositories based on
L<gitweb.cgi|https://git.wiki.kernel.org/index.php/Gitweb> and backed by
L<Catalyst>.

=head2 History

This project started off as an attempt to port I<gitweb.cgi> to a
Catalyst app in a piecemeal fashion. As it turns out, thanks largely
to Florian Ragwitz's earlier effort, it was easier to use I<gitweb.cgi>
as a template for building a new Catalyst application.

=head1 GETTING GITALIST

You can install Gitalist from CPAN in the usual way:

    cpan -i Gitalist

Alternatively, you can get Gitalist using git.

The canonical repository for the master branch is:

    git://git.shadowcat.co.uk/catagits/Gitalist.git

Gitalist is also mirrored to GitHub at L<https://github.com/broquaint/Gitalist>,
and a number of people have active forks
with branches and/or new features in the master branch.

=head1 BOOTSTRAPPING

As of C<0.002001> Gitalist can now be bootstrapped to run out of its
own directory by installing its prerequisites locally with the help of
L<local::lib>. So instead of installing the prerequisites to the
system path with CPAN they are installed under the Gitalist directory.

To do this clone Gitalist from the L<Shadowcat repository mentioned
above|/"GETTING GITALIST"> or grab a snapshot from broquaint's GitHub repository:

    https://github.com/broquaint/Gitalist/downloads

With the source acquired and unpacked run the following from within the
Gitalist directory:

    perl script/bootstrap.pl

This will install the necessary modules for the build process which in
turn installs the prerequisites locally.

B<NB:> The relevant bootstrap scripts aren't available in the CPAN dist
as the bootstrap scripts should not be installed.

=head1 INITIAL CONFIGURATION

Gitalist is configured using L<Catalyst::Plugin::Configloader>. The supplied sample
configuration is in L<Config::General> format, however it is possible to configure
Gitalist using other config file formats (such as YAML) if you prefer.

=head2 WHEN CHECKING GITALIST OUT OF GIT

Gitalist from git includes a minimal C<gitalist_local.conf>, which sets the repository
directory to one directory higher than the Gitalist repository.

This means that if you check Gitalist out next to your other git checkouts, then starting
the demo server needs no parameters at all:

    Gitalist [master]$ perl script/gitalist_server.pl
    You can connect to your server at http://localhost:3000

=head2 FOR CPAN INSTALLS

Gitalist can be supplied with a config file by setting the C<< GITALIST_CONFIG >>
environment variable to point to a configuration file.

If you install Gitalist from CPAN, a default configuration is installed along with gitalist,
which is complete except for a repository directory. You can get a copy of this configuration
by running:

  cp `perl -Ilib -MGitalist -e'print Gitalist->path_to("gitalist.conf")'` gitalist.conf

You can then edit this configuration, adding a C<repo_dir> path and customising
other settings as desired.

You can then start the Gitalist demo server by setting C<< GITALIST_CONFIG >>. For example:

    GITALIST_CONFIG=/usr/local/etc/gitalist.conf gitalist_server.pl

Alternatively, if you only want to set a repository directory and are otherwise happy with
the default configuration, then you can set the C<< GITALIST_REPO_DIR >> environment
variable, or pass the C<< --repo_dir >> flag to any of the scripts.

    GITALIST_REPO_DIR=/home/myuser/code/git gitalist_server.pl
    gitalist_server.pl --repo_dir home/myuser/code/git

The C<< GITALIST_REPO_DIR >> environment variable will override the repository directory set
in configuration, and will itself be overridden by he C<< --repo_dir >> flag.

=head1 RUNNING

Once you have followed the instructions above to install and configure Gitalist, you may want
to run it in a more production facing environment than using the single threaded developement
server.

The recommended deployment method for Gitalist is FastCGI, although Gitalist can also be run
under L<mod_perl|https://perl.apache.org/> or as pure Perl with L<Catalyst::Engine::PreFork>.

Assuming that you have installed Gitalist's dependencies into a L<local::lib>, and you
are running from a git checkout, adding a trivial FCGI script as C<script/gitalist.fcgi>
(this file is specifically in C<.gitignore> so you can have your own copy):

    #!/bin/sh
    exec /home/t0m/public_html/Gitalist/script/gitalist_fastcgi.pl

This example can be seen live here:

    http://example.gitalist.com

=head2 Plack

If you would like to run Gitalist under L<Plack> then one need only
make use of L<plackup|search.cpan.org/perldoc?plackup> and the
C<.psgi> found under C<scripts/>:

    plackup script/gitalist_app.psgi

=head1 CONFIGURATION

The Gitalist config is loaded with L<Catalyst::Plugin::ConfigLoader>
and the available config options are:

=over

=item no_gravatars

If true don't display gravatars. May be desirable if you are worried about
leaking repo names via the browser referer headers.

=back

=head2 Model::CollectionOfRepos

=over

=item repo_dir

A directory containing the directories to show.

=item repos

A list of directories containing repositories to show.

=item search_recursively

A boolean indicating whether to always search recursively for
repositories within C<repo_dir>.

=item whitelist

Path a file containing a list of repositories that can be shown. Each
line in the file will represent the name of a repo to show e.g

  Gitalist
  some-bare-repo.git

This is compatible with C<gitweb>'s C<projects.list>.

=item export_ok

If provided every must contain a file of the same name to be
visible. This is similar to C<gitweb>'s functionality.

=item class

If you want a different way of surfacing repositories you can use your
own model (i.e something that composes
L<Gitalist::Git::CollectionOfRepositories>) and specify the class name with
this config option.

=item args

Any additional arguments to be passed into the Model constructor, only
of use when used in conjunction with C<class> (see above).

=back

=head2 paging

=over

=item log

The number of commits to show in the summary, shortlog and longlog views.

=back

=head2 FASTCGI

Running Gitalist in FastCGI mode requires a webserver with FastCGI
support (such as apache with L<mod_fcgi|http://www.fastcgi.com/drupal/node/3>
or L<mod_fcgid|https://httpd.apache.org/mod_fcgid/>). Below is a sample
configuration using Apache2 with mod_fcgid in a dynamic configuration
(as opposed to static or standalone mode). More information on these modes and
their configuration can be found at L<Catalyst::Engine::FastCGI/"Standalone server mode">.

In Apache's F<mime.conf>, add C<AddHandler fcgid-script .fcgi>
(or C<AddHandler fastcgi-script .fcgi> for mod_fcgi).

And a quick VirtualHost configuration:

    <VirtualHost *:80>
        ServerName gitalist.yourdomain.com
        DocumentRoot /path/to/gitalist.fcgi
        <Directory "/path/to/gitalist.fcgi">
            AllowOverride all
            Options +ExecCGI -MultiViews +SymLinksIfOwnerMatch
            Order allow,deny
            Allow from all
        </Directory>

        # Tell Apache this is a FastCGI application
        <Files gitalist.fcgi>
            #change the below to fastcgi-script if using mod_fcgi
            SetHandler fcgid-script
        </Files>
    </VirtualHost>

Now to access your Gitalist instance, you'll go to
C<gitalist.yourdomain.com/gitalist.fcgi/> (B<do not forget that trailing> C</>).
If you'd like a different URL, of course, you'll likely want to use
L<mod_rewrite|https://httpd.apache.org/docs/mod/mod_rewrite.html> or equivalent.

If you find the need to do some troubleshooting, you can call
C<http://url_to_gitalist.fcgi?dump_info=1> and/or add export C<GITALIST_DEBUG=1>
to the top of your F<gitalist.fcgi> file (just below the shebang line).

=head2 Apache config

Apache will refuse C<%2F> in Gitalist URLs unless configured otherwise. Make
sure C<AllowEncodedSlashes On> is in your F<httpd.conf> file in order for
this to run smoothly.

To have the static content served statically by Apache, instead of
Gitalist, then add something like following line to your F<httpd.conf>:

    Alias /static  /usr/local/share/perl/5.10.1/Gitalist/root/static

=head1 CONTRIBUTING

Patches are welcome, please feel free to fork on github and send pull requests, send patches
from git format-patch to the bug tracker, or host your own copy of gitalist somewhere and
ask us to pull from it.

=head1 SUPPORT

Gitalist has an active irc community in C<#gitalist> on irc.perl.org, please feel free to stop
by and ask questions, report bugs or installation issues or generally for a chat about where
we plan to go with the project.

=head1 SEE ALSO

L<Gitalist::Controller::Root>

L<Git::Gitalist::Repository>

L<Catalyst>

=head1 AUTHORS AND COPYRIGHT

  Catalyst application:
    © 2009 Venda Ltd and Dan Brook <broq@cpan.org>
    © 2009, Tom Doran <bobtfish@bobtfish.net>
    © 2009, Zac Stevens <zts@cryptocracy.com>

  Original gitweb.cgi from which this was derived:
    © 2005-2006, Kay Sievers <kay.sievers@vrfy.org>
    © 2005, Christian Gierke

  Model based on http://github.com/rafl/gitweb
    © 2008, Florian Ragwitz

=head1 LICENSE

Licensed under GNU GPL v2
